### PR TITLE
hashtable.py: Changed hashtable

### PR DESCRIPTION
**Pull request by christianbender**
* I added a runner:  
```python
if __name__ == "__main__":
    unittest.main()
```  
* I simplified the code.

### CODE REVIEW

File: map/hashtable.py [hunk 1]
```diff
@@ -13,8 +14,8 @@ class HashTable(object):
     in Return True for a statement of the form key in map, if the given key is in the map, False otherwise.
     """
 
-    _empty = object()
-    _deleted = object()
+    _empty = None
+    _deleted = None
```
Comment (goswami-rahul) [inline map/hashtable.py]: `_empty` and `_deleted` act as placeholders for not yet assigned keys and deleted keys respectively. 
In many places, there are uses like  `self.keys[hash_] is _empty` and `self.keys[hash_] is _deleted` which won't work as expected if be replace them both by `None`.

**Comment by goswami-rahul** [comment]
I think the HashTable is designed such a way that you can only fill it with atmost `size` elements. If we delete a key from it, it is labeled as `_deleted` and no further key would be added to that hash. That means

File: map/hashtable.py (hunk 1 above)
Comment (christianbender) [reply map/hashtable.py]: @goswami-rahul  I ran all tests successful. After each change I ran the tests and it seems no problem.

**Comment by christianbender** [comment]
@goswami-rahul 
I ran all tests successful. After each change I ran the tests and it seems no problem.   
I now changed the tests and ran it successful.  
I changed  
```python
def test_delete_key(self):
```
---

```python
def test_changes_on_hash_table(self):
        m = HashTable(10)
        for i in range(10):
            m.put(i,i**2)
        print("\t --> "+str(m.get(2))) # DEBUG
        for i in range(10):
            self.assertEqual(i**2,m.get(i))
```
The test (above) fails.  ```m.get(0) = None```  

**important:** I guess the hash-function is the problem. If we use a better hash-function the problem can be solve. 

At the moment:  
```python 
def hash(self, key):
        return key % self.size
```

File: map/hashtable.py [hunk 2]
```diff
@@ -26,15 +27,14 @@ def put(self, key, value):
         initial_hash = hash_ = self.hash(key)
 
         while True:
-            if self._keys[hash_] is self._empty or self._keys[hash_] is self._deleted:
+            if not self._keys[hash_]:
```
Comment (goswami-rahul) [inline map/hashtable.py]: I found the issue with `m.get(0) == None`. Here this line should be reverted to the original version or `if self._keys[hash_] is None:`

**Comment by goswami-rahul** [comment]
`if not self._keys[hash_]:` confused also evaluated True for 0.

File: map/hashtable.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def put(self, key, value):
     def get(self, key):
         initial_hash = hash_ = self.hash(key)
         while True:
-            if self._keys[hash_] is self._empty:
+            if not self._keys[hash_]:
```
Comment (goswami-rahul) [inline map/hashtable.py]: revert back here too.

File: map/hashtable.py [hunk 4]
```diff
@@ -62,7 +62,7 @@ def get(self, key):
     def del_(self, key):
         initial_hash = hash_ = self.hash(key)
         while True:
-            if self._keys[hash_] is self._empty:
+            if not self._keys[hash_]:
```
Comment (goswami-rahul) [inline map/hashtable.py]: here

File: map/hashtable.py [hunk 5]
```diff
@@ -118,7 +118,7 @@ def __resize(self):
         self._keys = [self._empty] * self.size
         self._values = [self._empty] * self.size
         for key, value in zip(keys, values):
-            if key is not self._empty and key is not self._deleted:
+            if key:
```
Comment (goswami-rahul) [inline map/hashtable.py]: and here.

**Comment by goswami-rahul** [comment]
I suggest that we keep the extra tests that you implemented and revert back `_empty` to `object()` as it makes more sense to have the placeholder of empty as a unique object, not even `None`. After this we can also make anything as the value in hashtable (including `None`).

**Comment by christianbender** [comment]
@goswami-rahul  OK, I will see to it.

**Pull request closed without merging**